### PR TITLE
Keystone Trust added in Extensions

### PR DESCRIPTION
**Pull request by codevulture**
@jrperritt : Hi, as per discussion on: https://github.com/rackspace/gophercloud/pull/580 i have modified the changes, Please let me know your review points and suggestions.Thanks.

### CODE REVIEW

File: auth_options.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ type AuthOptions struct {
 	// possible.  This setting defaults to false.
 	//
 	// NOTE: The reauth function will try to re-authenticate endlessly if left unchecked.
-	// The way to limit the number of attempts is to provide a custom HTTP client to the provider client
```
Comment (yuanying) [inline auth_options.go]: Seems unneeded change.

File: openstack/client.go [hunk 2]
```diff
@@ -139,6 +139,7 @@ func v3auth(client *gophercloud.ProviderClient, endpoint string, options gopherc
 	v3Options := options
 
 	var scope *tokens3.Scope
+
```
Comment (yuanying) [inline openstack/client.go]: ditto

File: openstack/endpoint_location.go [hunk 3]
```diff
@@ -57,35 +57,35 @@ func V2EndpointURL(catalog *tokens2.ServiceCatalog, opts gophercloud.EndpointOpt
 // need to specify a Name and/or a Region depending on what's available on your OpenStack
 // deployment.
 func V3EndpointURL(catalog *tokens3.ServiceCatalog, opts gophercloud.EndpointOpts) (string, error) {
-	// Extract Endpoints from the catalog entries that match the requested Type, Interface,
```
Comment (yuanying) [inline openstack/endpoint_location.go]: ditto

File: openstack/identity/v3/extensions/auth_options.go [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+*/
+type AuthOptions struct {
+	//Populate fields in gophercloud AuthOptions also.
+	*gophercloud.AuthOptions
+
+	// Trust allows users to authenticate with Trust ID,
+	// The TrustID  field is to be used with Identity V3 API only.
+	// ID of the Trust.
+	TrustID string
+}
```
Comment (yuanying) [inline openstack/identity/v3/extensions/auth_options.go]: I guess this auth_options.go should be extensions/trust/auth_options.go
Because this option is only used by trust extension.

File: openstack/identity/v3/extensions/doc.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+// Package extensions provides information and interaction with the
+// different extensions available for the OpenStack Identity v3 service.
+package extensions
```
Comment (yuanying) [inline openstack/identity/v3/extensions/doc.go]: ditto

File: openstack/identity/v3/extensions/tokens/endpoint_location.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package tokens
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/endpoint_location.go]: I'm not sure why is this file needed?

File: openstack/identity/v3/extensions/tokens/errors.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
+package tokens
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/errors.go]: ditto, why could we re-use v3/tokens/errors.go ?

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
+package tokens
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: I guess token request for trust should be in extensions/trust/token.
And also this file seems contain redundant code.

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+	"github.com/rackspace/gophercloud/openstack/identity/v3/extensions"
+)
+
+// Scope allows a created token to be limited to a specific domain or project.
+type Scope struct {
+	ProjectID   string
+	ProjectName string
+	DomainID    string
+	DomainName  string
+	TrustID     string
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: If this request is only used by trust extension, ProjectID/ProjectName/DomainID/DomainName scope is not needed.

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+		"X-Subject-Token": subjectToken,
+	}
+}
+
+// Create authenticates and either generates a new token, or changes the Scope of an existing token.
+func Create(c *gophercloud.ServiceClient, options extensions.AuthOptions, scope *Scope) CreateResult {
+	type domainReq struct {
+		ID   *string `json:"id,omitempty"`
+		Name *string `json:"name,omitempty"`
+	}
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: If this `Create` method is only used for requesting trust token, domainReq is not needed.

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+
+	type projectReq struct {
+		Domain *domainReq `json:"domain,omitempty"`
+		Name   *string    `json:"name,omitempty"`
+		ID     *string    `json:"id,omitempty"`
+	}
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: ditto

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+
+	type v3Trust struct {
+		ID *string `json:"id,omitempty"`
+	}
+
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: Why is not this name `v3TrustReq` ?

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+	type identityReq struct {
+		Methods  []string     `json:"methods"`
+		Password *passwordReq `json:"password,omitempty"`
+		Token    *tokenReq    `json:"token,omitempty"`
+	}
+
+	type scopeReq struct {
+		Domain  *domainReq  `json:"domain,omitempty"`
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: Is this scope really needed?

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+		Project *projectReq `json:"project,omitempty"`
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: ditto

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+	if options.APIKey != "" {
+		return createErr(ErrAPIKeyProvided)
+	}
+	if options.TenantID != "" {
+		return createErr(ErrTenantIDProvided)
+	}
+	if options.TenantName != "" {
+		return createErr(ErrTenantNameProvided)
+	}
+
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: Following code is same as v3/tokens/requests.go. Could we re-use code?

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+			}
+
+			// Configure the request for UserID and Password authentication.
+			req.Auth.Identity.Password = &passwordReq{
+				User: userReq{ID: &options.UserID, Password: options.Password},
+			}
+		}
+	}
+
+	// Add a "scope" element if a Scope has been provided.
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: May be this code is only called by requesting trust token, so scope will surely has scope.TrustID.

File: openstack/identity/v3/extensions/trust/request.go [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	if endpoint != "" {
+		v3Client.Endpoint = endpoint
+	}
+
+	// copy the auth options to a local variable that we can change. `options`
+	// needs to stay as-is for reauth purposes
+	v3Options := options
+
+	var scope *trusttokens3.Scope
+
```
Comment (yuanying) [inline openstack/identity/v3/extensions/trust/request.go]: I think if options has no TrustID, we must delegate this request to `v3auth` method.

**Comment by codevulture** [comment]
@yuanying : Indentation and line changes are resultant of `go fmt` command which i ran on code , as this was the standard procedure on golang for codeformating like pep in python. So i think this could be ignores or please let me know if needed to be removed.

File: auth_options.go (hunk 1 above)
Comment (codevulture) [inline auth_options.go]: `go fmt` result.

File: openstack/client.go (hunk 2 above)
Comment (codevulture) [inline openstack/client.go]: `go fmt` result.

File: openstack/endpoint_location.go (hunk 3 above)
Comment (codevulture) [inline openstack/endpoint_location.go]: `go fmt` result.

File: openstack/identity/v3/extensions/auth_options.go (hunk 4 above)
Comment (codevulture) [inline openstack/identity/v3/extensions/auth_options.go]: I made it generic in extensions because if other extensions are implemented they could use it from here, but i could move it if only trust uses it ,Np.Thanks.

File: openstack/identity/v3/extensions/tokens/endpoint_location.go (hunk 6 above)
Comment (codevulture) [inline openstack/identity/v3/extensions/tokens/endpoint_location.go]: function V3EndpointURL(catalog *tokens3.ServiceCatalog, opts gophercloud.EndpointOpts) could not be used from `openstack/endpoint_location.go` so another method is required for that with different method signature, but implementing it in `openstack/endpoint_location.go` was not suitable so i moved it here.

File: openstack/identity/v3/extensions/tokens/errors.go (hunk 7 above)
Comment (codevulture) [inline openstack/identity/v3/extensions/tokens/errors.go]: yes, they could be reused, i thought it would be more neat!, but yes seems reasonable.

File: openstack/identity/v3/extensions/tokens/requests.go [hunk 18]
```diff
@@ -114,69 +86,10 @@ func Create(c *gophercloud.ServiceClient, options extensions.AuthOptions, scope
 			}
 		} else {
 			// If no password or token ID are available, authentication can't continue.
-			return createErr(ErrMissingPassword)
+			return createErr(tokens.ErrMissingPassword)
 		}
 	} else {
-		// Password authentication.
```
Comment (yuanying) [inline openstack/identity/v3/extensions/tokens/requests.go]: Why did you remove password authentication? We can use this if we provide trustID.

File: openstack/identity/v3/tokens/errors.go [hunk 19]
```diff
@@ -69,4 +69,7 @@ var (
 
 	// ErrScopeEmpty indicates that no credentials were provided in a Scope.
 	ErrScopeEmpty = errors.New("You must provide either a Project or Domain in a Scope")
+
+	// ErrScopeEmpty indicates that no credentials were provided in a Scope.
+	ErrTrustScope = errors.New("You must not provide password when authenticating with Trust")
```
Comment (yuanying) [inline openstack/identity/v3/tokens/errors.go]: You should add this error to extensions/trust/errors.go

**Comment by codevulture** [comment]
@jrperritt : Please Review the changes.

**Comment by codevulture** [comment]
@jrperritt : ping.

**Comment by jrperritt** [comment]
The `go fmt` doesn't bother me; in fact, we should be linting and checking for that on build. What bothers me is all of the copy/paste; that shouldn't be necessary. You should convert the `opts` function parameter to an `interface` and then have the original `AuthOpts` and your `AuthOpts` (`AuthOpts` plus TrustID) both separately implement the method required by the `interface`.

**Comment by codevulture** [comment]
@jrperritt :Thanks for comments, I would like to know some more details on how implementations needs to be proceeded to be clear about what needs to be done to have these merged, I have some questions that i would like to ask as i am new to gopherhead library:

I have these files below which have the major code in them, as per your suggestions which files should be changed/removed  and if possible can you specify exactly which things, that would help me a lot to implement the changes. like you specified  `opts` but i am not sure which function you referred, please provide your guidance and feedback:

 - openstack/identity/v3/extensions/auth_options.go
 - openstack/identity/v3/extensions/tokens/endpoint_location.go 
 - openstack/identity/v3/extensions/tokens/requests.go
 - openstack/identity/v3/extensions/tokens/requests_test.go
 - openstack/identity/v3/extensions/tokens/results.go
 - openstack/identity/v3/extensions/tokens/urls.go 
 - openstack/identity/v3/extensions/trust/request.go

**Comment by jrperritt** [comment]
There is no reason to have 2 open pull requests for the same feature. Discussion continued in #580
